wandb/sdk/launch/runner/vertex_runner.py

```diff
@@ -104,114 +96,95 @@ def run(
+            max_env_length=MAX_ENV_LENGTHS[self.__class__.__name__],
         )
 
-        wandb.termlog(
-            f"{LOG_PREFIX}Running training job {gcp_training_job_name} on {gcp_machine_type}."
-        )
+        worker_specs = spec_args.get("worker_pool_specs", [])
+        if not worker_specs:
+            raise LaunchError(
+                "Vertex requires at least one worker pool spec. Please specify "
+                "a worker pool spec in resource arguments under the key "
+                "`vertex.spec.worker_pool_specs`."
+            )
 
+        # TODO: Add entrypoint + args to each worker pool spec
+        for spec in worker_specs:
+            if not spec.get("container_spec"):
+                raise LaunchError(
+                    "Vertex requires a container spec for each worker pool spec. "
+                    "Please specify a container spec in resource arguments under "
+                    "the key `vertex.spec.worker_pool_specs[].container_spec`."
+                )
+            spec["container_spec"]["command"] = entry_cmd
+            spec["container_spec"]["env"] = [
+                {"name": k, "value": v} for k, v in env_vars.items()
+            ]
+
+        if not spec_args.get("staging_bucket"):
+            raise LaunchError(
+                "Vertex requires a staging bucket. Please specify a staging bucket "
+                "in resource arguments under the key `vertex.spec.staging_bucket`."
+            )
+        try:
+            aiplatform.init(
+                project=self.environment.project,
+                location=self.environment.region,
+                staging_bucket=spec_args.get("staging_bucket"),
+                credentials=self.environment.get_credentials(),
+            )
+            job = aiplatform.CustomJob(
+                display_name=launch_project.name,
+                worker_pool_specs=spec_args.get("worker_pool_specs"),
+                base_output_dir=spec_args.get("base_output_dir"),
+                encryption_spec_key_name=spec_args.get("encryption_spec_key_name"),
+                labels=spec_args.get("labels", {}),
+            )
+            execution_kwargs = dict(
+                timeout=run_args.get("timeout"),
+                service_account=run_args.get("service_account"),
+                network=run_args.get("network"),
+                enable_web_access=run_args.get("enable_web_access", False),
+                experiment=run_args.get("experiment"),
+                experiment_run=run_args.get("experiment_run"),
+                tensorboard=run_args.get("tensorboard"),
+                restart_job_on_worker_restart=run_args.get(
+                    "restart_job_on_worker_restart", False
+                ),
+            )
+        # Unclear if there are exceptions that can be thrown where we should
+        # retry instead of erroring. For now, just catch all exceptions and they
+        # go to the UI for the user to interpret.
```

Comment:
Didn't find any such errors and also can't get a list from GCP docs of what errors to expect. GPT4 made some suggestions but they all seemed like clear failure cases due to invalid arguments.